### PR TITLE
McAfee Web Gateway parser

### PR DESCRIPTION
**Pull request by squioc**
Add the module for McAfee and the parser for logs from McAfee Web Gateway

### CODE REVIEW

File: McAfee/mcafee_web_gateway/tests/mcafee_access_log.json [hunk 1]
```diff
@@ -0,0 +1,43 @@
+{
+  "input": {
+    "sekoiaio": {
+      "intake": {
+        "dialect": "mcafee-web-gateway",
+        "dialect_uuid": "40bac399-2d8e-40e3-af3b-f73a622c9687"
+      }
+    },
+    "message": "[06/Feb/2020:14:24:53 +0100] \"(BypassAuthentication-UserAgent)\" 1.2.3.4 403 \"GET http://ctldl.windowsupdate.com/msdownload/update/v3/static/trustedr/en/disallowedcertstl.cab?2975230343cffee9 HTTP/1.1\" \"Software/Hardware\" \"Minimal Risk\" \"\" 3252 238 \"Microsoft-CryptoAPI/10.0\" \"\" \"200\" \"\""
```
Comment (LeoArurault) [inline McAfee/mcafee_web_gateway/tests/mcafee_access_log.json:9]: This test case fails on `['event']['action']` field added when running locally

**Review by LeoArurault — CHANGES_REQUESTED** [review]
Missing logo in \mcafee_web_gateway/_meta`

File: McAfee/mcafee_web_gateway/_meta/fields.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+mcafee.webgateway.ruleset.name:
+  name: mcafee.webgateway.ruleset.name
+  description: "The name of the ruleset matching the request"
+  type: keyword
+
+mcafee.webgateway.ruleset.name:
+  name: mcafee.webgateway.ruleset.name
+  description: "The name of the ruleset matching the request"
+  type: keyword
```
[the remark this answers is not in the document]
Comment (squioc) [reply McAfee/mcafee_web_gateway/_meta/fields.yml]: Nice shot. Fixed

File: McAfee/mcafee_web_gateway/tests/mcafee_access_log.json (hunk 1 above)
Comment (squioc) [reply McAfee/mcafee_web_gateway/tests/mcafee_access_log.json]: fixed

**Review by LeoArurault — APPROVED** [review]
lgtm but I'm not an ECS expert

**Comment by squioc** [comment]
The GA expect a png logo, not a svg one. Fixed

File: McAfee/mcafee_web_gateway/_meta/fields.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+mcafee.webgateway.viruses:
+  name: mcafee.webgateway.viruses
+  description: "A list of virus name"
+  type: text
+
+mcafee.webgateway.ruleset.name:
+  name: mcafee.webgateway.ruleset.name
+  description: "The name of the ruleset matching the request"
+  type: keyword
```
Comment (otetard) [inline McAfee/mcafee_web_gateway/_meta/fields.yml:34]: Couldn’t we use [`rule.ruleset`](https://www.elastic.co/guide/en/ecs/current/ecs-rule.html#field-rule-ruleset) instead?

Comment (squioc) [reply McAfee/mcafee_web_gateway/_meta/fields.yml]: I missed this ECS field

File: McAfee/mcafee_web_gateway/_meta/fields.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+mcafee.webgateway.block.id:
+  name: mcafee.webgateway.block.id
+  description: "The block reason id of the action"
+  type: number
+
+mcafee.webgateway.block.reason:
+  name: mcafee.webgateway.block.reason
+  description: "The block reason of the action"
+  type: text
```
[the remark this answers is not in the document]
Comment (squioc) [reply McAfee/mcafee_web_gateway/_meta/fields.yml]: you're right

File: McAfee/mcafee_web_gateway/_meta/fields.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+mcafee.webgateway.ruleset.name:
+  name: mcafee.webgateway.ruleset.name
+  description: "The name of the ruleset matching the request"
+  type: keyword
+
+mcafee.webgateway.block.id:
+  name: mcafee.webgateway.block.id
+  description: "The block reason id of the action"
+  type: number
```
[the remark this answers is not in the document]
Comment (squioc) [reply McAfee/mcafee_web_gateway/_meta/fields.yml]: From the documentation, each ID is associated to one reason.

**Review by LeoArurault — DISMISSED** [review]

**Comment by squioc** [comment]
:warning: Important change: I moved the representation of the logs from a textual format (forcing us to use grok expression to parse the log) to a key-value format (as we can control the format used to generate the logs; a configuration is supplied in the documentation).

File: McAfee/mcafee_web_gateway/tests/mcafee_access_log_blocked.json [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      }
+    },
+    "message": "date=\"2022-03-11T10:39:16.390Z\" hostname=\"mwgproxy\" username=\"\" source_ip=1.2.3.4 destination_ip=2.2.2.41 destination_host=\"www.forbiddensite.com\" http_status_code=403 media_type=\"\" source_bytes=131 destination_bytes=0 http_request_first_line=\"GET http://www.forbiddensite.com/ HTTP/1.1\" url_categories=\"Pornography\" url_reputation_string=\"Minimal Risk\" url_reputation_code=-28 ruleset_name=\"Default\" rule_name=\"Block URLs Whose Category Is in Category Blocklist for Default Groups\" block_id=10 block_reason=\"Blocked by URL filtering\" body_infected=false virus_names=\"\" body_modified=false application_reputation=\"Unverified\" application_name=\"forbiddenapp\" http_referer=\"\" user_agent=\"curl/7.77.0\"",
+    "source": {"ip": "1.2.3.4"},
+    "destination": {"ip": "2.2.2.41"},
+    "event": {
+        "action": "block",
+        "start": "2022-03-11T10:39:16.390Z",
+        "code": "10",
+        "reason": "Blocked by URL filtering"
```
Comment (otetard) [inline McAfee/mcafee_web_gateway/tests/mcafee_access_log_blocked.json:26]: You are missing an ECS categorization. For blocked URLs, you could use `event.type="denied"` and `event.category="network"`.

More information on ECS categorization fields: https://www.elastic.co/guide/en/ecs/current/ecs-category-field-values-reference.html

File: McAfee/mcafee_web_gateway/tests/mcafee_access_log_blocked.json [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        "ruleset": "Default",
+        "name": "Block URLs Whose Category Is in Category Blocklist for Default Groups"
+    },
+    "mcafee": {
+        "webgateway": {
+            "url": {
+                "reputation": "Minimal Risk",
+                "reputation_code": -28
+            },
+            "categories": ["Pornography"],
```
Comment (otetard) [inline McAfee/mcafee_web_gateway/tests/mcafee_access_log_blocked.json:53]: This information seems in the source log seems to be related to the URL itself. Maybe you could move this information to `mcafee.webgateway.url.categories`.

File: McAfee/mcafee_web_gateway/ingest/parser.yml [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            item_sep: \s
+  - name: parse_first_line
+    external:
+      name: grok.match
+      properties:
+        input_field: '{{parse_kv.message.http_request_first_line}}'
+        output_field: message
+        pattern: '%{WORD:http_method} %{URL:url} HTTP/%{NUMBER:http_version}'
+        custom_patterns:
+          REQUEST_LINE: '%{WORD:http_method} %{URL:url} HTTP/%{NUMBER:http_version}'
```
Comment (LeoArurault) [inline McAfee/mcafee_web_gateway/ingest/parser.yml:19]: Unused code it seems
```suggestion
```

Comment (squioc) [reply McAfee/mcafee_web_gateway/ingest/parser.yml]: fixed

File: McAfee/mcafee_web_gateway/tests/mcafee_access_log_blocked.json (hunk 7 above)
Comment (squioc) [reply McAfee/mcafee_web_gateway/tests/mcafee_access_log_blocked.json]: moved

File: McAfee/mcafee_web_gateway/tests/mcafee_access_log_blocked.json (hunk 6 above)
Comment (squioc) [reply McAfee/mcafee_web_gateway/tests/mcafee_access_log_blocked.json]: Fixed

**Review by LeoArurault — APPROVED** [review]

**Review by otetard — APPROVED** [review]

**Pull request merged**